src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -25,17 +25,404 @@ namespace XBU = XBUtilities;
 #include "core/common/system.h"
 #include "core/common/device.h"
 #include "core/common/error.h"
+#include "core/pcie/common/device_pcie.h"
+#include "core/common/query_requests.h"
+#include "core/common/error.h"
+#include "flash/flasher.h"
 
 // 3rd Party Library - Include Files
-#include <boost/format.hpp>
-#include <boost/program_options.hpp>
+#include "boost/format.hpp"
+#include "boost/tokenizer.hpp"
+#include "boost/filesystem.hpp"
+#include "boost/program_options.hpp"
 namespace po = boost::program_options;
 
 // System - Include Files
 #include <iostream>
 #include <fstream>
 #include <thread>
 #include <chrono>
+#include <ctime>
+
+#ifdef _WIN32
+#pragma warning(disable : 4996)
+#endif
+
+
+// =============================================================================
+
+// ------ L O C A L   F U N C T I O N S ---------------------------------------
+
+namespace {
+/*
+ * Update shell on the board
+ */
+static void 
+update_shell(uint16_t index, const std::string& flashType,
+    const std::string& primary, const std::string& secondary)
+{
+  std::unique_ptr<firmwareImage> pri;
+  std::unique_ptr<firmwareImage> sec;
+
+  if (!flashType.empty()) {
+    std::cout << "CAUTION: Overriding flash mode is not recommended. " <<
+      "You may damage your card with this option." << std::endl;
+  }
+
+  Flasher flasher(index);
+  if(!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % index));
+
+  if (primary.empty())
+    throw xrt_core::error("Shell not specified");
+
+  pri = std::make_unique<firmwareImage>(primary.c_str(), MCS_FIRMWARE_PRIMARY);
+  if (pri->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % primary.c_str()));
+  if (!secondary.empty()) {
+    sec = std::make_unique<firmwareImage>(secondary.c_str(),
+      MCS_FIRMWARE_SECONDARY);
+    if (sec->fail())
```

Comment:
No. This is how this works. You pass in a dsabin to both, primary and secondary, parameters. If the platform takes only 1 image, it will set the 2nd image to nullptr and "upgradeFirmware" calls the upgrade shell with 1 image appropriately. If it takes in 2 images then  "upgradeFirmware" calls upgrade with 2 images